Implement batch processing for aggregate functions with multiple nullable arguments

Algunenano opened the pull request:
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement batch processing for aggregate functions with multiple nullable arguments


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

I've noticed this when working on https://github.com/ClickHouse/ClickHouse/pull/40633. Functions like argMin, argMax, avgWeighted and so on won't use batch processing of data if they have nullable arguments. We can do it by merging the flags as done in multiple places already.

Some examples:

* `SELECT avgWeighted(toNullable(number), number) t FROM numbers(1000000000) SETTINGS max_threads=1`
  - 22.8: 3.780s
  - changes: 2.519s

* `SELECT avgWeighted(number, toNullable(number)) t FROM numbers(1000000000) SETTINGS max_threads=1`
  - 22.8: 3.857s
  - changes: 2.527s

* `SELECT avgWeighted(toNullable(number), toNullable(number)) t FROM numbers(1000000000) SETTINGS max_threads=1`
  - 22.8: 4.309s
  - changes: 2.483

* `SELECT avgWeightedIf(toNullable(number), toNullable(number), toNullable(number) % 10) t FROM numbers(1000000000) SETTINGS max_threads=1`
  - 22.8: 6.227s
  - changes: 3.373s

Algunenano committed:
Implement batch processing for aggregate functions with multiple nullable arguments

Algunenano committed:
Fix broken perf test

Algunenano commented:
There seems to be some performance changes in the `distinct_in_order` tests but I don't see how my changes could have affected that code directly. I'm guessing it might be a codegen change or maybe some interference in the host during the perf run?

nickitat commented on src/AggregateFunctions/AggregateFunctionNull.h:
```diff
@@ -414,6 +414,89 @@ class AggregateFunctionNullVariadic final
+        }
+        else
+        {
+            final_flags = std::make_unique<UInt8[]>(row_end);
```
maybe just OR filters 1, 2, ... into filter 0?

Algunenano replied:
Uhmm, I think that's only possible when if is provided since the other filters are const (which makes sense since they can come from anywhere, including a direct view to the disk I guess), but it makes sense to add that optimization when possible and reduce memory usage.

nickitat commented on src/AggregateFunctions/AggregateFunctionNull.h:
```diff
@@ -414,6 +414,89 @@ class AggregateFunctionNullVariadic final
+            final_flags_ptr = final_flags.get();
+            for (size_t i = row_begin; i < row_end; i++)
+            {
+                for (size_t f = 0; f < nullable_filters.size(); f++)
```
pls use `filter` instead of `f`

Algunenano replied:
👍 will do

nickitat commented on tests/performance/avg_weighted.xml:
```diff
@@ -32,5 +32,21 @@
     <query>SELECT avgWeighted(num_u, num)   FROM perf_avg FORMAT Null</query>
     <query>SELECT avgWeighted(num_u, num_u) FROM perf_avg FORMAT Null</query>
 
+    <query>SELECT avgWeighted(num_f,                num_f)             FROM perf_avg FORMAT Null</query>
```
changes in this pr are not particularly coupled with `avgWeighted`, so add pls a few tests with other aggregate functions

Algunenano replied:
The only agreggate function with multiple arguments (-if argument doesn't count) that I've seen implementing addMany* with a perf improvement was avgWeighted.

ArgMin and ArgMax will also benefit once #40633 is merged and I hope that in the future more aggregations (with one or more args) will implement custom batch functions since a lot of them can benefit from it in one way or the other.

Algunenano committed:
Improve filter handling in addBatchSinglePlace with nullable arguments

Algunenano commented:
Added some changes:
* Changed f -> filter
* AggregateFunctionNullVariadic::addBatchSinglePlace: Use a single auxiliar flag array (before it'd use 2 in some cases)
* Iterate per (filter, row) instead of (row, filter) when merging filters: This is much faster. As expected. No idea why I went the other way. I've updated the description with the new value for `SELECT avgWeighted(toNullable(number), toNullable(number)) t FROM numbers(1000000000) SETTINGS max_threads=1`.

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into null_variadic

Algunenano committed:
Fix detecting the Null filter usage

novikd commented on src/AggregateFunctions/AggregateFunctionNull.h:
```diff
@@ -414,6 +414,109 @@ class AggregateFunctionNullVariadic final
+
+            if (included_elements == 0)
+                return;
+            if (included_elements != (row_end - row_begin))
```
@Algunenano  This line looks quite strange. It seems `included_elements` is not supposed to be `bool`.

Algunenano replied:
Yes it seems wrong. I'll have a look, look for similar issues (similar code was added in other plac3s) see if it affects results and send a fix.

Alex-Cheng commented on src/AggregateFunctions/AggregateFunctionNull.h:
```diff
@@ -414,6 +414,109 @@ class AggregateFunctionNullVariadic final
+
+        if (if_argument_pos >= 0)
+        {
+            final_flags = std::make_unique<UInt8[]>(row_end);
```
why not create array with size of (row_end - row_begin)?

Algunenano replied:
Although it would save space it might cause problems in the called functions which expect a full array. They shouldn’t dereference the beginning of the array (it’d be a bug if they do), but passing a pointer to memory that can’t be derefenced (you’d need to pass ‘data() - row_begin’) is tricky and I’d only consider it if the improvement was noticeable.

Also, back then the usage of row_begin different than 0 was minimal. Probably close to 0. That changed recently

Alex-Cheng commented on src/AggregateFunctions/AggregateFunctionNull.h:
```diff
@@ -414,6 +414,109 @@ class AggregateFunctionNullVariadic final
+            const auto & flags = assert_cast<const ColumnUInt8 &>(*columns[if_argument_pos]).getData();
+            for (size_t i = row_begin; i < row_end; i++)
+            {
+                final_flags[i] = !flags.data()[i];
```
the index for accessing to array `final_flags` is between row_begin and row_end , I think the [0, row_begin) part of the array will never be accessed.
